### PR TITLE
DIT-559 Add Keyword Search

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/assets/javascripts/search.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+var advancedSearch = {
+    removeKeyword: function(index) {
```
Comment (ed0906) [inline app/assets/javascripts/search.js]: Afraid the only way we can do "remove keyword" as far as I'm aware is to use JS

File: app/uk/gov/hmrc/tariffclassificationfrontend/forms/SearchForm.scala [hunk 2]
```diff
@@ -19,22 +19,28 @@ package uk.gov.hmrc.tariffclassificationfrontend.forms
...
       "trader_name" -> optional(text.verifying(_ => true)),
       "commodity_code" -> optional(text.verifying(emptyOr(numeric, minLength(2), maxLength(22)): _*)),
-      "live_rulings_only" -> optional(boolean)
-    )(SearchFormData.apply)(SearchFormData.unapply)
+      "live_rulings_only" -> optional(boolean),
+      "keyword" -> optional(set(text))
+    )(Search.apply)(Search.unapply)
+  )
+
+  val formWithoutValidation: Form[Search] = Form(
```
Comment (ed0906) [inline app/uk/gov/hmrc/tariffclassificationfrontend/forms/SearchForm.scala]: Used in the Search Binder

File: app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala [hunk 3]
```diff
@@ -82,6 +82,7 @@ class BindingTariffClassificationConnector @Inject()(configuration: AppConfig, c
       .++(search.traderName.map(queryBinder.unbind("trader_name", _)))
       .++(search.commodityCode.map(queryBinder.unbind("commodity_code", _)))
       .++(search.liveRulingsOnly.filter(identity).map(_ => queryBinder.unbind("min_decision_end", Instant.now(clock).toString) + "&" + queryBinder.unbind("status", CaseStatus.COMPLETED.toString)))
+      .++(search.keywords.map(set => set.map(queryBinder.unbind("keyword", _)).mkString("&")))
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala]: `set => set` can be simplified as `_`

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SearchController.scala [hunk 4]
```diff
@@ -33,27 +33,33 @@ import scala.concurrent.Future.successful
...
           }
-        }
-      )
-
+        )
+      })
     }
   }
 
+
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SearchController.scala]: lot of empty lines

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Search.scala [hunk 5]
```diff
@@ -37,30 +38,25 @@ object Search {
...
-
-      Some(Right(Search(
-        traderName = param(traderNameKey),
-        commodityCode = param(commodityCodeKey).map(_.replaceAll(" ", "")),
-        liveRulingsOnly = param(liveRulingsOnlyKey).flatMap(bindBoolean)
-      )))
+    override def bind(string: String, requestParams: Map[String, Seq[String]]): Option[Either[String, Search]] = {
+      val filteredParams: Map[String, Seq[String]] = requestParams
+        .mapValues(_.map(_.trim).filter(_.nonEmpty))
+        .filter {
```
Comment (gokyo) [inline app/uk/gov/hmrc/tariffclassificationfrontend/models/Search.scala]: all this block can be simplified as 
```
    .filter(_._2.nonEmpty)
```

File: app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala (hunk 3 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/connector/BindingTariffClassificationConnector.scala]: Done

File: app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SearchController.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/controllers/SearchController.scala]: Sorted

File: app/uk/gov/hmrc/tariffclassificationfrontend/models/Search.scala (hunk 5 above)
Comment (ed0906) [reply app/uk/gov/hmrc/tariffclassificationfrontend/models/Search.scala]: Sorted

**Pull request merged**